src/OpenTelemetry.Instrumentation.AWSLambda/Implementation/AWSMessagingUtils.cs

```diff
@@ -0,0 +1,155 @@
+// <copyright file="AWSMessagingUtils.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Amazon.Lambda.SNSEvents;
+using Amazon.Lambda.SQSEvents;
+using Newtonsoft.Json;
+using OpenTelemetry.Context.Propagation;
+
+namespace OpenTelemetry.Instrumentation.AWSLambda.Implementation;
+
+internal class AWSMessagingUtils
+{
+    // SNS attribute types: https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html
+    private const string SnsAttributeTypeString = "String";
+    private const string SnsAttributeTypeStringArray = "String.Array";
+    private const string SnsMessageAttributes = "MessageAttributes";
+
+    internal static (PropagationContext ParentContext, IEnumerable<ActivityLink> Links) ExtractParentContext(SQSEvent sqsEvent)
+    {
+        if (sqsEvent?.Records == null)
+        {
+            return (default, null);
+        }
+
+        // We assume there can be only one parent that's why we consider only a single (the last) record as the carrier.
+        var parentRecord = sqsEvent.Records.LastOrDefault();
+        var parentContext = ExtractParentContext(parentRecord);
+
+        var links = new List<ActivityLink>();
+        foreach (var record in sqsEvent.Records)
+        {
+            var context = ExtractParentContext(record);
```

Comment:
I think we should add an optimization here for the case of `object.ReferenceEquals(record, parentRecord)` where we reuse the `parentContext` instead of extracting again.
Especially big improvement for the common case of a single record.

Reply:
Thanks. I added this optimisation.